apps/twap/src/app/plugins/infura.ts

```diff
@@ -0,0 +1,237 @@
+import { OrderParameters, SupportedChainId } from '@cowprotocol/cow-sdk';
+import { ethers } from 'ethers';
+import { FastifyInstance } from 'fastify';
+import {
+  ComposableCoW__factory,
+  ConditionalOrderCreatedEvent,
+} from '@cow-web-services/abis';
+import fp from 'fastify-plugin';
+import { getConditionalOrderId } from '../utils/getConditionalOrderId';
+import { getOrderStatus } from '../utils/getOrderStatus';
+import { parseOrderStruct } from '../utils/parseOrderStruct';
+import { generateOrderParts } from '../utils/generateOrderParts';
+import { Wallet } from '../data/wallet';
+import { Order } from '../data/order';
+import { OrderPart } from '../data/orderPart';
+import { Block } from '../data/block';
+
+// Add lag per network (in confirmations)
+const NETWORKS = ['mainnet', 'goerli'];
+const CONFIRMATIONS: Record<SupportedChainId, number> = {
+  '1': 5,
+  '5': 3,
+  '100': 3,
+};
+
+export const COMPOSABLE_COW_ADDRESS: Record<SupportedChainId, string> = {
+  1: '0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738',
+  100: '0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738',
+  5: '0xF487887DA5a4b4e3eC114FDAd97dc0F785d72738',
+};
+
+export interface TwapPartOrderItem {
+  uid: string;
+  index: number;
+  chainId: SupportedChainId;
+  safeAddress: string;
+  twapOrderId: string;
+  order: OrderParameters;
+}
+
+type SafeTransactionParams = {
+  submissionDate: string;
+  executionDate: string | null;
+  isExecuted: boolean;
+  nonce: number;
+  confirmationsRequired: number;
+  confirmations: number;
+  safeTxHash: string;
+};
+
+interface ConditionalOrderParams {
+  staticInput: string;
+  salt: string;
+  handler: string;
+}
+
+export interface TwapOrdersSafeData {
+  conditionalOrderParams: ConditionalOrderParams;
+  safeTxParams: SafeTransactionParams;
+}
+
+function infuraPlugin(
```

Comment:
This was a bit hard to follow. 
We could add some more comments explaining what is happening.

Reply:
Good one, I'll update this. I should be able to extract the async function in here.